### PR TITLE
feat: Add Fluent String helpers

### PR DESCRIPTION
**Pull request by kkumar-gcc**
RelatedTo : https://github.com/goravel/goravel/issues/82

## ✅ Checks

- [x] My pull request adheres to the code style of this project
- [x] My code requires changes to the documentation
- [ ] I have updated the documentation as required
- [ ] All the tests have passed

### CODE REVIEW

File: support/str/str.go
Comment (hwbrzzl) [inline support/str/str.go:1]: Perfect, please add a complete test.

File: support/str/str_test.go [hunk 1]
```diff
@@ -4,11 +4,1036 @@ import (
...
+func (s *StringTestSuite) TestMatchAll() {
+	s.Equal([]string{"World"}, Of("Hello, World!").MatchAll("World"))
+	s.Equal([]string{"(test)"}, Of("This is a (test) string").MatchAll(`\([^)]+\)`))
+	s.Equal([]string{"123", "456"}, Of("abc123def456def").MatchAll(`\d+`))
+	s.Equal([]string(nil), Of("No match here").MatchAll(`\d+`))
+	s.Equal([]string{"Hello, World!"}, Of("Hello, World!").MatchAll(""))
+	s.Equal([]string{"[456]"}, Of("123 [456]").MatchAll(`\[456\]`))
+}
+
+func (s *StringTestSuite) TestIsMatch() {
```
Comment (hwbrzzl) [inline support/str/str_test.go:335]: You said that you use Laravel's test cases, but there are some differences. Is there a specific reason, please?

image

File: support/str/str_test.go [hunk 2]
```diff
@@ -4,11 +4,1036 @@ import (
...
+	s.Equal("Fooar", Of("Foobar").Remove("f", "b").String())
+	s.Equal("Foobar", Of("Foo|bar").Remove("f", "|").String())
+}
+
+func (s *StringTestSuite) TestRepeat() {
+	s.Equal("aaaaa", Of("a").Repeat(5).String())
+	s.Equal("", Of("").Repeat(5).String())
+}
+
+func (s *StringTestSuite) TestReplace() {
```
Comment (hwbrzzl) [inline support/str/str_test.go:404]: Ditto

File: support/str/str_test.go [hunk 3]
```diff
@@ -4,11 +4,1036 @@ import (
...
+	s.Equal("Goravel framework", Of("goravel framework").UcFirst().String())
+}
+
+func (s *StringTestSuite) TestUcSplit() {
+	s.Equal([]string{"Krishan", "Kumar"}, Of("KrishanKumar").UcSplit())
+	s.Equal([]string{"Hello", "From", "Goravel"}, Of("HelloFromGoravel").UcSplit())
+	s.Equal([]string{"He_llo_", "World"}, Of("He_llo_World").UcSplit())
+}
+
+func (s *StringTestSuite) TestUnless() {
```
Comment (hwbrzzl) [inline support/str/str_test.go:586]: Missing

**Review by hwbrzzl — COMMENTED** [review]
A perfect PR 👍

File: support/str/str.go [hunk 4]
```diff
@@ -3,11 +3,966 @@ package str
...
+	if end > strLen {
+		end = strLen
+	}
+
+	// Return the substring.
+	return string(runes[start:end])
+}
+
+// Max returns the largest of x or y.
+func Max[T constraints.Ordered](x T, y T) T {
```
[the remark this answers is not in the document]
Comment (kkumar-gcc) [reply support/str/str.go:959]: if i change it to max it collides with the 'builtin' function which will be supported in `golang 1.21`.

File: support/str/str.go [hunk 5]
```diff
@@ -3,11 +3,966 @@ package str
...
+		return s
+	}
+
+	s.value = strings.Join(words[:limit], " ") + defaultEnd
+	return s
+}
+
+// FieldsFunc splits the input string into words with preservation, following the rules defined by
+// the provided functions f and preserveFunc.
+func FieldsFunc(s string, f func(rune) bool, preserveFunc ...func(rune) bool) []string {
```
[the remark this answers is not in the document]
Comment (kkumar-gcc) [reply support/str/str.go:874]: okay

File: support/str/str_test.go (hunk 1 above)
Comment (kkumar-gcc) [reply support/str/str_test.go:335]: Both are the same. It's just that regex is defined a little differently than PHP.

File: support/str/str.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (hwbrzzl) [reply support/str/str.go:959]: Emm, we may choose a new name.

File: support/str/str_test.go (hunk 3 above)
Comment (kkumar-gcc) [reply support/str/str_test.go:586]: Sorry, missed it

File: support/str/str.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (kkumar-gcc) [reply support/str/str.go:959]: Okay

**Review by hwbrzzl — COMMENTED** [review]
Just left to optimize the function import, and then the PR can be merged.

**Review by hwbrzzl — APPROVED** [review]
A great PR!

**Pull request merged**